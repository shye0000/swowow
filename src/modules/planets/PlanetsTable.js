import React from 'react';
import {Link} from 'react-router-dom';
import {connect} from 'react-redux';
import HomeLink from '../../components/HomeLink';
import {fetchPlanets, setCurrentPlanet} from './redux/actions';
import { slowDown, speedUp } from '../../redux/actions';
import Table from 'antd/lib/table';
import './PlanetsTable.scss';

class PlanetsTable extends React.Component {

	componentDidMount() {
		this.props.fetchPlanets();
	}

	pagination = {
		current: 1
	}

	columns = [{
		title: 'Name',
		dataIndex: 'name',
		key: 'name',
		render: (text, record) => {
			const planetUrl = encodeURIComponent(record.url);
			return <Link
				to={`/planets/${planetUrl}`}
				onClick={() => this.props.setCurrentPlanet(record)}
			>{text}</Link>;
		}
	}, {
		title: 'Diameter',
		dataIndex: 'diameter',
		key: 'diameter',
	}, {
		title: 'Climate',
		dataIndex: 'climate',
		key: 'climate',
	}, {
		title: 'Population',
		dataIndex: 'population',
		key: 'population',
	}];

	addKeyToDataSource = (dataSource) => {
		if (!dataSource || !dataSource.length) {
			return undefined;
		}
		return dataSource.map(record => ({
			...record,
			key: record.url,
		}));
	}

	getTablePaginationByCount = (count) => {
		return {
			...this.pagination,
			total: count
		};
	}

	handleTableOnChange = (pagination) => {
		if (pagination.current !== this.pagination.current) {
			this.pagination.current = pagination.current;
			this.props.fetchPlanets(pagination.current);
		}
	}

	render () {
		const {fetching, planetsCollection} = this.props;
		let dataSource, tablePagination;
		if (planetsCollection) {
			const {results, count} = planetsCollection;
			dataSource = this.addKeyToDataSource(results);
			tablePagination = this.getTablePaginationByCount(count);
		}
		return <div className="planets-table">
			<HomeLink />
			<div className="table-name">STAR WARS PLANETS</div>
			<Table
				className="table-body" loading={fetching}
				columns={this.columns} pagination={tablePagination} dataSource={dataSource}
				onChange={(pagination) => this.handleTableOnChange(pagination)}
			/>
		</div>;
	}
}

const mapStateToProps = (state) => {
	return {
		fetching: state.planets.fetching,
		fetchSuccess: state.planets.fetchSuccess,
		planetsCollection: state.planets.planetsCollection
	};
};

const mapDispatchToProps = (dispatch) => {
	return {
		setCurrentPlanet: (planet) => {
			dispatch(setCurrentPlanet(planet));
		},
		fetchPlanets: (page) => {
			dispatch(speedUp());
			dispatch(fetchPlanets(page)).then(() => dispatch(slowDown())).catch(() => dispatch(slowDown()));
		}
	};
};

export default connect(mapStateToProps, mapDispatchToProps)(PlanetsTable);