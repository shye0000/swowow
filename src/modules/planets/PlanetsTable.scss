@import "../../styles/vars";
.planets-table {
  height: 100%;
  padding: $sm-padding $lg-padding;
  overflow-y: auto;
  overflow-x: hidden;
  color: white;
  font-weight: bold;
  .table-name {
    margin-top: $lg-padding;
    font-size: 2em;
  }
  .table-body {
    background: none;
    margin-top: $lg-padding;
    .ant-table-placeholder {
      background: none;
    }
    th, td {
      background: none;
      color: white;
    }
    tr:hover {
      td {
        background: rgba(60,60,60,0.35);
        color: white;
      }
    }
  }
  .ant-pagination-prev, .ant-pagination-next {
    a {
      background: none;
      border-style: dashed;
      border-width: 2px;
      color: white;
      font-weight: bold;
    }
  }

  .ant-pagination-item {
    border-style: dashed;
    border-width: 2px;
    background: none;
    &:not(.ant-pagination-item-active) {
      a {
        color: white;
      }
    }
  }
}