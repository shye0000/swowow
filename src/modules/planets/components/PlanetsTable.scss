@import "../../../styles/vars";
.planets-table {
  height: 100%;
  padding: $sm-padding $lg-padding;
  overflow-y: auto;
  overflow-x: hidden;
  color: white;
  font-weight: bold;
  .table-name {
    margin-top: $lg-padding;
    font-size: 2em;
  }
  .table-body {
    background: none;
    margin-top: $lg-padding;
    .ant-table-placeholder {
      background: none;
    }
    th, td {
      background: none;
      color: white;
    }
    tr:hover {
      td {
        background: rgba(60,60,60,0.35);
        color: white;
      }
    }
  }
  .ant-pagination-prev, .ant-pagination-next {
    margin-bottom: 10px;
    a {
      background: none;
      border-style: dashed;
      border-width: 2px;
      color: white;
      font-weight: bold;
    }
    &:hover {
      a {
        color: white;
      }
    }
    &.ant-pagination-disabled {
      opacity: 0.5;
    }
  }

  .ant-pagination-item {
    margin-bottom: 10px;
    border-style: dashed;
    border-width: 2px;
    background: none;
    font-weight: bold;

    &:not(.ant-pagination-item-active) {
      a {
        color: white;
      }
    }
  }
}