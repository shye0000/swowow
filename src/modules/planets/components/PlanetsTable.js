import React, {useEffect, useState} from 'react';
import {Link} from 'react-router-dom';
import {connect} from 'react-redux';
import HomeLink from '../../../components/HomeLink';
import {fetchPlanets, setCurrentPlanet} from '../redux/actions';
import { slowDown, speedUp } from '../../../redux/actions';
import Table from 'antd/lib/table';
import ReactResizeDetector from 'react-resize-detector';
import './PlanetsTable.scss';

const columns = [{
	title: 'Name',
	dataIndex: 'name',
	key: 'name',
	width: 100,
	render(text, record){
		const planetUrl = encodeURIComponent(record.url);
		return <Link
			to={`/planets/${planetUrl}`}
			onClick={() => setCurrentPlanet(record)}
		>{text}</Link>;
	},
}, {
	title: 'Diameter',
	dataIndex: 'diameter',
	key: 'diameter',
	width: 100,
}, {
	title: 'Climate',
	dataIndex: 'climate',
	key: 'climate',
	width: 200,
}, {
	title: 'Population',
	dataIndex: 'population',
	key: 'population',
	width: 200
}];

const addKeyToDataSource = (dataSource) => {
	if (!dataSource || !dataSource.length) {
		return undefined;
	}
	return dataSource.map(record => ({
		...record,
		key: record.url,
	}));
};

const PlanetsTable = props => {
	const [width, setWidth] = useState(null);

	const {
		fetchPlanets,
		fetching,
		planetsCollection,
		currentPage,
	} = props;

	useEffect(() => {
		if (!fetching && !planetsCollection) {
			fetchPlanets();
		}
	});

	const handleTableOnChange = (pag) => {
		if (pag.current !== currentPage) {
			fetchPlanets(pag.current);
		}
	};

	let dataSource, tablePagination;
	if (planetsCollection) {
		const {results, total_records} = planetsCollection;
		dataSource = addKeyToDataSource(results);
		tablePagination = {
			current: currentPage,
			total: total_records,
		};
	}

	return <div className="planets-table">
		<HomeLink />
		<div className="table-name">STAR WARS PLANETS</div>
		<Table
			className="table-body"
			scroll={{x: width < 600 ? 600 : undefined}}
			loading={fetching}
			columns={columns}
			pagination={tablePagination}
			dataSource={dataSource}
			onChange={(pag) => handleTableOnChange(pag)}
		/>
		<ReactResizeDetector handleWidth onResize={(width) => setWidth(width)} />
	</div>;
};

const mapStateToProps = (state) => {
	return {
		fetching: state.planets.fetching,
		fetchSuccess: state.planets.fetchSuccess,
		planetsCollection: state.planets.planetsCollection,
		currentPage: state.planets.currentPage,
	};
};

const mapDispatchToProps = (dispatch) => {
	return {
		fetchPlanets: (page) => {
			dispatch(speedUp());
			dispatch(fetchPlanets(page)).then(() => dispatch(slowDown())).catch(() => dispatch(slowDown()));
		}
	};
};

export default connect(mapStateToProps, mapDispatchToProps)(PlanetsTable);