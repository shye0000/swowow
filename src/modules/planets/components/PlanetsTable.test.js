import React from 'react';
import { PlanetsTable } from './PlanetsTable';
import { shallow } from 'enzyme';

describe('PlanetsTable', () => {

	it.skip('render correctly', () => {
		const props = {
			fetching: false,
			fetchSuccess: true,
			planetsCollection: {},
			fetchPlanets: jest.fn(() => {})
		};
		const wrapper = shallow(<PlanetsTable {...props} />);
		expect(wrapper).toMatchSnapshot();
	});

	it.skip('fetchPlanets get te be called after first rendering', () => {
		const props = {
			fetching: false,
			fetchSuccess: true,
			planetsCollection: {},
			fetchPlanets: jest.fn(() => {})
		};
		const wrapper = shallow(<PlanetsTable {...props} />);
		const componentDidMountSpy = jest.spyOn(wrapper.instance(), 'componentDidMount');
		componentDidMountSpy();
		expect(props.fetchPlanets).toBeCalled();
	});

	it.skip('fetchPlanets get te be called when table change', () => {
		const props = {
			fetching: false,
			fetchSuccess: true,
			planetsCollection: {},
			fetchPlanets: jest.fn(() => {})
		};
		const wrapper = shallow(<PlanetsTable {...props} />);
		const handleTableOnChangeSpy = jest.spyOn(wrapper.instance(), 'handleTableOnChange');
		wrapper.instance().pagination.current = 1;
		handleTableOnChangeSpy({current: 3});
		expect(props.fetchPlanets).toBeCalled();
	});

	// todo
	// it('fetchPlanets should not be called when table change but with the same pagination', () => {
	// 	const props = {
	// 		fetching: false,
	// 		fetchSuccess: true,
	// 		planetsCollection: {},
	// 		fetchPlanets: jest.fn(() => {})
	// 	};
	// 	const wrapper = shallow(<PlanetsTable {...props} />);
	// 	const handleTableOnChangeSpy = jest.spyOn(wrapper.instance(), 'handleTableOnChange');
	// 	wrapper.instance().pagination.current = 3;
	// 	handleTableOnChangeSpy({current: 3});
	// 	expect(props.fetchPlanets).not.toBeCalled();
	// });

	it.skip('addKeyToDataSource get to be called with results of planetCollection only when render with planetCollection in props', () => {
		const props = {
			fetching: false,
			fetchSuccess: false,
			fetchPlanets: jest.fn(() => {})
		};
		const wrapper = shallow(<PlanetsTable {...props} />);
		const addKeyToDataSourceSpy = jest.spyOn(wrapper.instance(), 'addKeyToDataSource');
		wrapper.setProps({
			fetching: false,
			fetchSuccess: false,
			fetchPlanets: jest.fn(() => {})
		});
		expect(addKeyToDataSourceSpy).not.toBeCalled();
		wrapper.setProps({
			fetching: false,
			fetchSuccess: true,
			planetsCollection: {results: []},
			fetchPlanets: jest.fn(() => {})
		});
		expect(addKeyToDataSourceSpy).toBeCalledWith([]);
	});

	it.skip('getTablePaginationByCount get to be called only when render with planetCollection in props', () => {
		const props = {
			fetching: false,
			fetchSuccess: false,
			fetchPlanets: jest.fn(() => {})
		};
		const wrapper = shallow(<PlanetsTable {...props} />);
		const getTablePaginationByCountSpy = jest.spyOn(wrapper.instance(), 'getTablePaginationByCount');
		wrapper.setProps({
			fetching: false,
			fetchSuccess: false,
			fetchPlanets: jest.fn(() => {})
		});
		expect(getTablePaginationByCountSpy).not.toBeCalled();
		wrapper.setProps({
			fetching: false,
			fetchSuccess: true,
			planetsCollection: {},
			fetchPlanets: jest.fn(() => {})
		});
		expect(getTablePaginationByCountSpy).toBeCalled();
	});

});