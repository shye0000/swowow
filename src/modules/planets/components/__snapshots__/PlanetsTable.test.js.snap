// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlanetsTable render correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <PlanetsTable
    fetchPlanets={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
      }
    }
    fetchSuccess={true}
    fetching={false}
    planetsCollection={Object {}}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <HomeLink />,
        <div
          className="table-name"
        >
          STAR WARS PLANETS
        </div>,
        <Table
          bordered={false}
          className="table-body"
          columns={
            Array [
              Object {
                "dataIndex": "name",
                "key": "name",
                "render": [Function],
                "title": "Name",
                "width": 100,
              },
              Object {
                "dataIndex": "diameter",
                "key": "diameter",
                "title": "Diameter",
                "width": 100,
              },
              Object {
                "dataIndex": "climate",
                "key": "climate",
                "title": "Climate",
                "width": 200,
              },
              Object {
                "dataIndex": "population",
                "key": "population",
                "title": "Population",
                "width": 200,
              },
            ]
          }
          dataSource={Array []}
          indentSize={20}
          loading={false}
          locale={Object {}}
          onChange={[Function]}
          pagination={
            Object {
              "current": 1,
              "total": undefined,
            }
          }
          prefixCls="ant-table"
          rowKey="key"
          showHeader={true}
          size="large"
          useFixedHeader={false}
        />,
      ],
      "className": "planets-table",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "STAR WARS PLANETS",
          "className": "table-name",
        },
        "ref": null,
        "rendered": "STAR WARS PLANETS",
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "bordered": false,
          "className": "table-body",
          "columns": Array [
            Object {
              "dataIndex": "name",
              "key": "name",
              "render": [Function],
              "title": "Name",
              "width": 100,
            },
            Object {
              "dataIndex": "diameter",
              "key": "diameter",
              "title": "Diameter",
              "width": 100,
            },
            Object {
              "dataIndex": "climate",
              "key": "climate",
              "title": "Climate",
              "width": 200,
            },
            Object {
              "dataIndex": "population",
              "key": "population",
              "title": "Population",
              "width": 200,
            },
          ],
          "dataSource": Array [],
          "indentSize": 20,
          "loading": false,
          "locale": Object {},
          "onChange": [Function],
          "pagination": Object {
            "current": 1,
            "total": undefined,
          },
          "prefixCls": "ant-table",
          "rowKey": "key",
          "showHeader": true,
          "size": "large",
          "useFixedHeader": false,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <HomeLink />,
          <div
            className="table-name"
          >
            STAR WARS PLANETS
          </div>,
          <Table
            bordered={false}
            className="table-body"
            columns={
              Array [
                Object {
                  "dataIndex": "name",
                  "key": "name",
                  "render": [Function],
                  "title": "Name",
                  "width": 100,
                },
                Object {
                  "dataIndex": "diameter",
                  "key": "diameter",
                  "title": "Diameter",
                  "width": 100,
                },
                Object {
                  "dataIndex": "climate",
                  "key": "climate",
                  "title": "Climate",
                  "width": 200,
                },
                Object {
                  "dataIndex": "population",
                  "key": "population",
                  "title": "Population",
                  "width": 200,
                },
              ]
            }
            dataSource={Array []}
            indentSize={20}
            loading={false}
            locale={Object {}}
            onChange={[Function]}
            pagination={
              Object {
                "current": 1,
                "total": undefined,
              }
            }
            prefixCls="ant-table"
            rowKey="key"
            showHeader={true}
            size="large"
            useFixedHeader={false}
          />,
        ],
        "className": "planets-table",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "STAR WARS PLANETS",
            "className": "table-name",
          },
          "ref": null,
          "rendered": "STAR WARS PLANETS",
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "bordered": false,
            "className": "table-body",
            "columns": Array [
              Object {
                "dataIndex": "name",
                "key": "name",
                "render": [Function],
                "title": "Name",
                "width": 100,
              },
              Object {
                "dataIndex": "diameter",
                "key": "diameter",
                "title": "Diameter",
                "width": 100,
              },
              Object {
                "dataIndex": "climate",
                "key": "climate",
                "title": "Climate",
                "width": 200,
              },
              Object {
                "dataIndex": "population",
                "key": "population",
                "title": "Population",
                "width": 200,
              },
            ],
            "dataSource": Array [],
            "indentSize": 20,
            "loading": false,
            "locale": Object {},
            "onChange": [Function],
            "pagination": Object {
              "current": 1,
              "total": undefined,
            },
            "prefixCls": "ant-table",
            "rowKey": "key",
            "showHeader": true,
            "size": "large",
            "useFixedHeader": false,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
